src/v/storage/chunk_cache.cc

```diff
@@ -54,16 +86,25 @@ ss::future<chunk_cache::chunk_ptr> chunk_cache::get() {
     if (!_sem.waiters()) {
         return do_get();
     }
-    return ss::get_units(_sem, 1).then(
-      [this](ssx::semaphore_units) { return do_get(); });
+
+    return wait_and_get();
 }
 
 ss::future<chunk_cache::chunk_ptr> chunk_cache::do_get() {
     if (auto c = pop_or_allocate(); c) {
         return ss::make_ready_future<chunk_ptr>(c);
     }
-    return ss::get_units(_sem, 1).then(
-      [this](ssx::semaphore_units) { return do_get(); });
+
+    return wait_and_get();
+}
+
+ss::future<chunk_cache::chunk_ptr> chunk_cache::wait_and_get() {
+    auto fut = ss::get_units(_sem, 1);
+    if (_sem.waiters()) {
+        _wait_for_chunk_count++;
```

Comment:
> i wonder if the counter should be bumped if fut is not available, rather than if there are waiters.

So I would kinda expect those two ways to be (almost) the same? We use the waiters pattern in a different place (being blocked on the kafka memory semaphore) but I feel like both should do the same?

Reply:
```
get_units   waiters    blocks     counter bump
ready             0        no              no
(A) ready        >0        no     eek. didn't block, but bumped counter
(B) not ready     0        yes    eek. blocked, but didn't bump counter
not ready        >0        yes             yes

(A) impossible because it would imply the caller skipped the FIFO queue
(B) impossible because the caller will be counted in the waiters count
```

yeh i think this makes sense. i was confused because it seemed like there would be scenarios where the counting wouldn't reflect what was really happening (A, B) but I think i've convinced myself that those cases aren't possible. maybe a bit less clear (at least to me) than directly checking if the current fiber will block or not on the future.

Reply:
> i wonder if the chunk cache even serves a useful purpose that couldn't be replaced by a simple semaphore for memory usage. 

I mostly tend to agree that the chunk cache is not necessary. Allocation is cheap compared to the costs involved with processing one chunk (and regardless we do a bunch of allocations on the write path anyway). I suppose it does reduce fragmentation in some sense by keeping all the chunks in a contiguous region, though at 16K they (just) fit in the small pool which I think also helps a lot with fragmentation.

Reply:
FWIW I went though this same thought process when I ran into the `if (_sem.waiters())` pattern elsewhere. I also concluded they are identical, though I find `f.available()` clearer (and occasionally handy since if `available()` you can call `get()` immediately and dispense with co_await/then).

Reply:
Also, I think this future will _never_ be available? As far as I understand, this semaphore is always zero: it starts zero and the only way it is signaled is when it has waiters, which means it stays zero (since the waiter immediately consumes the signal, it never visibly increments).

Reply:
>Also, I think this future will never be available?

Ah, good point I do think this is true. It's a bit convoluted, but the original goal was to only invoke the semaphore machinery (the counting, the linked lists of waiters, waking up, etc...) only when necessary. given how cheap it is to allocate a chunk, we didn't want that accounting to negate the benefits of the cache. but it was all motivated by vibes and not measurements, so we ended up with something difficult to analyze and maybe unnecessary :/

Reply:
@dotnwat thanks for the background, it makes a lot of sense.